packages/core/src/invoke.ts

```diff
@@ -2,7 +2,7 @@
 /** @author Brian Cavalier */
 /** @author John Hann */
 
-export default function invoke (f, args) {
+export default function invoke <F extends (...args: any[]) => any>(f: F, args: Parameters<F>): ReturnType<F> {
```

Comment:
Maybe `<F extends (...args: unknown[]) => unknown>`?

Reply:
:+1: I'm in favor, if it works.

Reply:
Unfortunately it doesn't work because all calls to `invoke` should define function using `unknown`s as well because `...args` are in contravariant position so TS checks type assignments vice versa. `unknown` in return also doesn't type check because it cannot be assign to `ReturnType<F>`. Event if I replace `ReturnType` with custom conditional type it still doesn't work.

Reply:
It works only this way:
```typescript
export default function invoke <Args extends unknown[], R> (f: (...args: Args) => R, args: Args): R {
  /* eslint complexity: [2,7] */
  switch (args.length) {
    case 0: {
      return (f as () => R)()
    }
    case 1: {
      return (f as unknown as (a: unknown) => R)(args[0])
    }
    case 2: return (f as unknown as (a: unknown, b: unknown) => R)(args[0], args[1])
    case 3: return (f as unknown as (a: unknown, b: unknown, c: unknown) => R)(args[0], args[1], args[2])
    case 4: return (f as unknown as (a: unknown, b: unknown, c: unknown, d: unknown) => R)(args[0], args[1], args[2], args[3])
    case 5: return (f as unknown as (a: unknown, b: unknown, c: unknown, d: unknown, e: unknown) => R)(args[0], args[1], args[2], args[3], args[4])
    default:
      return f.apply(undefined, args)
  }
}
```

Reply:
Should I commit it? There's no `any` in function signature and all casts are inside its body

Reply:
> It works only this way:
> 
> ```ts
> export default function invoke <Args extends unknown[], R> (f: (...args: Args) => R, args: Args): R {
>   /* eslint complexity: [2,7] */
>   switch (args.length) {
>     case 0: {
>       return (f as () => R)()
>     }
>     case 1: {
>       return (f as unknown as (a: unknown) => R)(args[0])
>     }
>     case 2: return (f as unknown as (a: unknown, b: unknown) => R)(args[0], args[1])
>     case 3: return (f as unknown as (a: unknown, b: unknown, c: unknown) => R)(args[0], args[1], args[2])
>     case 4: return (f as unknown as (a: unknown, b: unknown, c: unknown, d: unknown) => R)(args[0], args[1], args[2], args[3])
>     case 5: return (f as unknown as (a: unknown, b: unknown, c: unknown, d: unknown, e: unknown) => R)(args[0], args[1], args[2], args[3], args[4])
>     default:
>       return f.apply(undefined, args)
>   }
> }
> ```

_Phooo!_ That’s not nice. In that case, I would favor the `any` with a comment to look into how types good be bettered. Maybe it requires a broader thinking.

Reply:
ok, I left a comment